### PR TITLE
WT-8771 Checkpoint cleanup to dirty obsolete page with overflow items

### PR DESCRIPTION
**Pull request by kommiharibabu**

### CODE REVIEW

**Review by quchenhao — APPROVED** [review]
lgtm

File: src/btree/bt_sync.c [hunk 1]
```diff
@@ -147,12 +147,14 @@ __sync_ref_list_pop(WT_SESSION_IMPL *session, WT_REF_LIST *rlp, uint32_t flags)
...
-         * let them clean during reconciliation.
+         * Mark the obsolete page dirty to let the reconciliation to remove the updates from page,
+         * except if the page is already marked as empty by previous reconciliation. The obsolete
+         * pages with overflow keys cannot be fast deleted, so marking them dirty will let them
+         * clean during reconciliation.
          */
         WT_RET(__wt_page_modify_init(session, rlp->list[i]->page));
-        __wt_page_modify_set(session, rlp->list[i]->page);
+        if (rlp->list[i]->page->modify->rec_result != WT_PM_REC_EMPTY)
+            __wt_page_modify_set(session, rlp->list[i]->page);
```
Comment (keithbostic) [inline src/btree/bt_sync.c:157]: The phrase "let them clean during reconciliation" was confusing? Maybe reword for clarity?

**Review by keithbostic — APPROVED** [review]
lgtm

File: src/btree/bt_sync.c (hunk 1 above)
Comment (kommiharibabu) [reply src/btree/bt_sync.c]: Fixed.

**Review by quchenhao — APPROVED** [review]
lgtm

File: src/btree/bt_sync.c [hunk 2]
```diff
@@ -147,25 +160,26 @@ __sync_ref_list_pop(WT_SESSION_IMPL *session, WT_REF_LIST *rlp, uint32_t flags)
...
+            WT_RET(__wt_page_modify_init(session, rlp->list[i].ref->page));
+            __wt_page_modify_set(session, rlp->list[i].ref->page);
+        }
 
         /*
          * Ignore the failure from urgent eviction. The failed refs are taken care in the next
          * checkpoint.
          */
-        WT_IGNORE_RET_BOOL(__wt_page_evict_urgent(session, rlp->list[i]));
+        WT_IGNORE_RET_BOOL(__wt_page_evict_urgent(session, rlp->list[i].ref));
```
Comment (keithbostic) [inline src/btree/bt_sync.c:175]: Why is this eviction urgent?

**Review by keithbostic — APPROVED** [review]
lgtm

File: src/btree/bt_sync.c (hunk 2 above)
Comment (kommiharibabu) [reply src/btree/bt_sync.c:175]: I don't think it needs an urgent eviction. Other than it clears the cache when the page is removed sooner. And also I feel we don't have any proper exposed API to add a page to the current eviction queue.

File: src/btree/bt_sync.c [hunk 3]
```diff
@@ -9,10 +9,18 @@
...
 typedef struct {
-    WT_REF **list;
+    WT_REF *ref;
+    bool ovfl_items; /* Contains overflow items? */
+} WT_OBSOLETE_REF;
+
+/*
+ * A list of obsolete WT_REF's.
+ */
+typedef struct {
```
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/btree/bt_sync.c:22]: I agree that having a flag in ref to indicate whether it has any overflow items may be useful. Already there is a flag in addr type to indicate whether it is an overflow page or not? Adding another flag may confuse.

File: src/btree/bt_sync.c (hunk 2 above)
Comment (keithbostic) [reply src/btree/bt_sync.c:175]: Makes sense -- all we can really do is set the page generation low enough so that hopefully the page gets picked sooner. My concern here is if we flood the urgent queue with these pages, pages we really might need to urgently evict might get lost in the noise. I'll leave it up to you if you think that's a valid concern -- if we really don't care when these pages make it out, setting the page generation seems sufficient.

File: src/btree/bt_sync.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (keithbostic) [reply src/btree/bt_sync.c:22]: I agree: what you have is good.

File: src/btree/bt_sync.c (hunk 2 above)
Comment (kommiharibabu) [reply src/btree/bt_sync.c]: I agree with your concern. As you suggested, I changed the code to set the page read generation number to a lower value, so that it will be picked for eviction sooner instead of checkpoint cleanup forcing it to evict.

**Comment by quchenhao** [comment]
@kommiharibabu I don't understand the new change. Where do you set the read generation number?

**Comment by kommiharibabu** [comment]
@quchenhao Earlier as part of checkpoint cleanup we add the obsolete pages into the urgent eviction queue. It is possible that due to these pages getting added to the urgent queue, the pages that are really required for the urgent eviction may be ignored. So instead of we add it to the queue, set the read generation number to a lower value, so these pages are picked by the eviction server sooner to evict.

**Review by quchenhao — APPROVED** [review]
lgtm

File: src/btree/bt_sync.c [hunk 4]
```diff
@@ -317,14 +261,28 @@ __sync_ref_obsolete_check(WT_SESSION_IMPL *session, WT_REF *ref, WT_REF_LIST *rl
...
-        hazard = false;
+        /*
+         * Dirty the obsolete page with overflow items to let the page reconciliation remove all the
+         * overflow items.
+         */
+        if (ovfl_items) {
+            WT_ERR(__wt_page_modify_init(session, ref->page));
+            __wt_page_modify_set(session, ref->page);
+        }
+
```
Comment (quchenhao) [inline src/btree/bt_sync.c:283]: May be add a comment here.

Comment (keithbostic) [reply src/btree/bt_sync.c:283]: Maybe just move the comment at line 271 to line 284?

**Review by keithbostic — APPROVED** [review]
I didn't think about the fact that you would no longer need an array of WT_REFs if you bumped the page generation instead of using the urgent queue. I like this change a lot, nice!!

File: src/btree/bt_sync.c (hunk 4 above)
Comment (kommiharibabu) [reply src/btree/bt_sync.c:279]: Fixed.
